ovn_k8s/modes/overlay.py

```diff
@@ -436,3 +466,366 @@ def sync_services(self, services):
+            rule_address_set = self._build_rule_address_set_name(
+                policy_data['metadata']['name'],
+                namespace,
+                rule)
+            ovn_nbctl('--if-exists', 'remove', 'address_set',
+                      rule_address_set, 'addresses', pod_ip)
+
+    def build_pseudo_acls(self, policy_data):
+        # Build pseudo ACL list for policy
+        policy_id = policy_data['metadata']['uid']
+        policy_pseudo_acls = []
+        for rule in policy_data['spec']['ingress']:
+            ports_data = rule.get('ports', [])
+            protocol_ports = []
+            for item in ports_data:
+                protocol_ports.append((item['protocol'], item.get('port')))
+            from_data = rule.get('from')
+            if not from_data:
+                src_pod_ips = '*'
+            else:
+                rule_address_set = self._build_rule_address_set_name(
+                    policy_data['metadata']['name'],
+                    policy_data['metadata']['namespace'],
+                    rule)
+                src_pod_ips = 'address_set(%s)' % rule_address_set
+            pseudo_acl = (POLICY_ACL_PRIORITY,
+                          protocol_ports, src_pod_ips,
+                          'allow-related')
+            policy_pseudo_acls.append(pseudo_acl)
+        self.pseudo_acls[policy_id] = policy_pseudo_acls
+        return policy_pseudo_acls
+
+    def remove_pseudo_acls(self, policy_id):
+        self.pseudo_acls.pop(policy_id, None)
+
+    def _find_policies_for_pod(self, pod_data, policies):
+        pod_labels = pod_data['metadata'].get('labels', {})
+        pod_policies = []
+        for policy in policies:
+            policy_spec = policy['spec']
+            pod_selector = policy_spec.get('podSelector', {}).get(
+                'matchLabels', {})
+            # TODO(salv-orlando): Implement not only equality based selectors
+            if pod_selector:
+                for label in set(pod_labels.keys()) & set(pod_selector.keys()):
+                    if pod_labels[label] == pod_selector[label]:
+                        pod_policies.append(policy)
+                        # policy matched, move on
+                        break
+            else:
+                # An empty selector matches all pods in the namespace
+                pod_policies.append(policy)
+        return pod_policies
+
+    def apply_pod_policy_acls(self, pod_data, policies):
+        namespace = pod_data['metadata']['namespace']
+        pod_name = pod_data['metadata']['name']
+        pod_policies = self._find_policies_for_pod(pod_data, policies)
+        # Always start with a drop all rule
+        pseudo_acls = [(DEFAULT_ACL_PRIORITY, [], '*', 'drop')]
+        for policy in pod_policies:
+            policy_pseudo_acls = self.pseudo_acls.get(
+                policy['metadata']['uid'])
+            if not policy_pseudo_acls:
+                policy_pseudo_acls = self.build_pseudo_acls(policy)
```

Comment:
The address set is actually created for every rule in the policy.
There is a bug actually, and that bug is that the address set is created also when it's not needed. I will double check your comment however as I would not be surprised if you found another bug.

Otherwise I will add inline comments to clarify how address sets are created.